File: sandbox/math/vector3.cpp
#include "vector3.hpp"

#include <glm/glm.hpp>
#include <glm/vec3.hpp>

namespace sbx {

inline glm::vec3 to_glm_vector3(const vector3& v) {
  return glm::vec3(v.x(), v.y(), v.z());
}

inline vector3 to_sbx_vector3(const glm::vec3& v) {
  return vector3(v.x, v.y, v.z);
}

vector3::vector3()
: _x{0.0f},
  _y{0.0f},
  _z{0.0f} { }

vector3::vector3(const float32 x, const float32 y, const float32 z)
: _x{x},
  _y{y},
  _z{z} { }

vector3 vector3::cross(const vector3& lhs, const vector3& rhs) {
  const auto cross = glm::cross(to_glm_vector3(lhs), to_glm_vector3(rhs));
  return to_sbx_vector3(cross);
}

float32 dot(const vector3& lhs, const vector3& rhs) {
  return glm::dot(to_glm_vector3(lhs), to_glm_vector3(rhs));
}

constexpr bool vector3::operator==(const vector3& other) const {
  return _x == other._x && _y == other._y && _z == other._z;
}

constexpr bool vector3::operator!=(const vector3& other) const {
  return !(*this == other);
}

vector3& vector3::operator+=(const vector3& other) {
  _x += other._x;
  _y += other._y;
  _z += other._z;

  return *this;
}

vector3& vector3::operator-=(const vector3& other) {
  _x -= other._x;
  _y -= other._y;
  _z -= other._z;

  return *this;
}

vector3& vector3::operator*=(const float32 scalar) {
  _x *= scalar;
  _y *= scalar;
  _z *= scalar;

  return *this;
}

vector3& vector3::operator/=(const float32 scalar) {
  _x /= scalar;
  _y /= scalar;
  _z /= scalar;

  return *this;
}

vector3 vector3::operator-() const {
  return vector3(-_x, -_y, -_z);
}

vector3& vector3::operator-() {
  _x = -_x;
  _y = -_y;
  _z = -_z;

  return *this;
}

float32 vector3::x() const {
  return _x;
}

float32 vector3::y() const {
  return _y;
}

float32 vector3::z() const {
  return _z;
}

float32& vector3::x() {
  return _x;
}

float32& vector3::y() {
  return _y;
}

float32& vector3::z() {
  return _z;
}

vector2 vector3::xy() const {
  return vector2(_x, _y);
}

vector2 vector3::xz() const {
  return vector2(_x, _z);
}

vector2 vector3::yz() const {
  return vector2(_y, _z);
}

float32 vector3::length() const {
  return glm::length(to_glm_vector3(*this));
}

void vector3::normalize() {
  const auto normalized = glm::normalize(to_glm_vector3(*this));
  *this = to_sbx_vector3(normalized);
}

vector3 vector3::normalized() const {
  const auto normalized = glm::normalize(to_glm_vector3(*this));
  return to_sbx_vector3(normalized);
}

vector3 operator+(vector3 lhs, const vector3& rhs) {
  lhs += rhs;
  return lhs;
}

vector3 operator-(vector3 lhs, const vector3& rhs) {
  lhs -= rhs;
  return lhs;
}

vector3 operator*(vector3 lhs, const float32 rhs) {
  lhs *= rhs;
  return lhs;
}

vector3 operator/(vector3 lhs, const float32 rhs) {
  lhs /= rhs;
  return lhs;
}

} // namespace sbx

File: dotnet/Sbx.Core/Math/Vector3.cs
using System.Runtime.InteropServices;

namespace Sbx.Core.Math
{

  [StructLayout(LayoutKind.Sequential, Pack = 4)]
  public struct Vector3 : IEquatable<Vector3>
  {

    public static Vector3 Zero = new Vector3(0, 0, 0);
    public static Vector3 One = new Vector3(1, 1, 1);
    public static Vector3 Forward = new Vector3(0, 0, -1);
    public static Vector3 Back = new Vector3(0, 0, 1);
    public static Vector3 Right = new Vector3(1, 0, 0);
    public static Vector3 Left = new Vector3(-1, 0, 0);
    public static Vector3 Up = new Vector3(0, 1, 0);
    public static Vector3 Down = new Vector3(0, -1, 0);

    public static Vector3 Inifinity = new Vector3(float.PositiveInfinity);

    public float X;
    public float Y;
    public float Z;

    public Vector3(float x, float y, float z)
    {
      X = x;
      Y = y;
      Z = z;
    }

    public Vector3(float scalar)
    : this(scalar, scalar, scalar)
    { }

    public Vector3(float x, Vector2 yz)
    : this(x, yz.X, yz.Y)
    { }

    public Vector3(Vector2 xy, float z = 0.0f)
    : this(xy.X, xy.Y, z)
    { }

    public static Vector3 operator*(Vector3 left, float scalar)
    {
      return new Vector3(left.X * scalar, left.Y * scalar, left.Z * scalar);
    }

    public static Vector3 operator*(float scalar, Vector3 right)
    {
      return new Vector3(scalar * right.X, scalar * right.Y, scalar * right.Z);
    }

    public static Vector3 operator*(Vector3 left, Vector3 right)
    {
      return new Vector3(left.X * right.X, left.Y * right.Y, left.Z * right.Z);
    }

		public static Vector3 operator/(Vector3 left, Vector3 right) 
    {
      return new Vector3(left.X / right.X, left.Y / right.Y, left.Z / right.Z);
    }
    
		public static Vector3 operator/(Vector3 left, float scalar) 
    {
      return new Vector3(left.X / scalar, left.Y / scalar, left.Z / scalar);
    }
    
		public static Vector3 operator/(float scalar, Vector3 right) 
    {
      return new Vector3(scalar/ right.X, scalar/ right.Y, scalar/ right.Z);
    }
    
		public static Vector3 operator+(Vector3 left, Vector3 right) 
    {
      return new Vector3(left.X + right.X, left.Y + right.Y, left.Z + right.Z);
    }
    
		public static Vector3 operator+(Vector3 left, float right) 
    {
      return new Vector3(left.X + right, left.Y + right, left.Z + right);
    }
    
		public static Vector3 operator-(Vector3 left, Vector3 right) 
    {
      return new Vector3(left.X - right.X, left.Y - right.Y, left.Z - right.Z);
    }
    
		public static Vector3 operator-(Vector3 left, float right) 
    {
      return new Vector3(left.X - right, left.Y - right, left.Z - right);
    }

    public static Vector3 operator-(Vector3 vector)
    {
      return new Vector3(-vector.X, -vector.Y, -vector.Z);
    }

    public override bool Equals(object obj)
    {
      return obj is Vector3 other && Equals(other);
    }

    public bool Equals(Vector3 other) {
      return X == other.X && Y == other.Y && Z == other.Z;
    }

    public override int GetHashCode()
    {
      return (X, Y, Z).GetHashCode();
    }

    public static bool operator==(Vector3 left, Vector3 right)
    {
      return left.Equals(right);
    }

    public static bool operator!=(Vector3 left, Vector3 right)
    {
      return !(left == right);
    }

    public float LengthSquared()
    {
      return X * X + Y * Y + Z * Z;
    }

    public float Length()
    {
      return MathF.Sqrt(LengthSquared());
    }

    public Vector3 Normalized()
		{
			float length = Length();

			return length > 0.0f ? Create(this, v => v / length) : new Vector3(X, Y, Z);
		}

		public void Normalize()
		{
			float length = Length();

			if (length > 0.0f)
      {
				Apply(v => v / length);
      }
		}

    public float Distance(Vector3 other)
		{
			return MathF.Sqrt(MathF.Pow(other.X - X, 2) + MathF.Pow(other.Y - Y, 2) + MathF.Pow(other.Z - Z, 2));
		}

		public static float Distance(Vector3 lhs, Vector3 rhs)
		{
			return MathF.Sqrt(MathF.Pow(rhs.X - lhs.X, 2) + MathF.Pow(rhs.Y - lhs.Y, 2) + MathF.Pow(rhs.Z - lhs.Z, 2));
		}

    public static float Dot(Vector3 a, Vector3 b)
    {
      return a.X * b.X + a.Y * b.Y + a.Z * b.Z;
    }

    public static Vector3 Cross(Vector3 a, Vector3 b)
		{
			return new Vector3(
				a.Y * b.Z - b.Y * a.Z,
				a.Z * b.X - b.Z * a.X,
				a.X * b.Y - b.X * a.Y
			);
		}

    public static Vector3 Abs(Vector3 vector)
    {
      return Create(vector, MathF.Abs);
    }

    public static Vector3 Clamp(Vector3 vector, Vector3 min, Vector3 max)
    {
      return new Vector3(
        System.Math.Clamp(vector.X, min.X, max.X),
        System.Math.Clamp(vector.Y, min.Y, max.Y),
        System.Math.Clamp(vector.Z, min.Z, max.Z)
      );
    }

    public static Vector3 Cos(Vector3 vector)
    {
      return Create(vector, MathF.Cos);
    }

    public static Vector3 Sin(Vector3 vector)
    {
      return Create(vector, MathF.Sin);
    }

    public static Vector3 Lerp(Vector3 a, Vector3 b, float t)
    {
      t = t < 0.0f ? 0.0f : (t > 1.0f ? 1.0f : t);

      return a + (b - a) * t;
    }

    public override string ToString()
    {
      return string.Format("Vector3[{0}, {1}, {2}]", X, Y, Z);
    }

    public void Apply(Func<float, float> function)
		{
			X = function(X);
			Y = function(Y);
			Z = function(Z);
		}

    public static Vector3 Create(Vector3 vector, Func<float, float> function)
    {
      return new Vector3(function(vector.X), function(vector.Y), function(vector.Z));
    }

  } // struct Vector3

} // namespace Sbx.Core.Math
